### PR TITLE
internal team communication form

### PR DESCRIPTION
**Pull request by casey1173**
Adds a communication model for assessments and studies for internal, team-member only discussions about ongoing assessments.
### Assessment Communications
Project managers can update assessment communications under Assessment Editing in the Actions button.

image

This opens a form for editing communications

image

Assessment communications are visible on the main assessment page, at the bottom. These are only visible for team members and higher.

image

### Study Communications
Team members can update study communications under Study editing, in the Actions button.

image

Study communications are visible on the main study page. Again, these communications are only visible for team members.

image

### CODE REVIEW

**Comment by casey1173** [comment]
Adds a communication model for assessments and studies for internal, team-member only discussions about ongoing assessments.
### Assessment Communications
Project managers can update assessment communications under Assessment Editing in the Actions button.

image


This opens a form for editing communications

image


Assessment communications are visible on the main assessment page, at the bottom. These are only visible for team members and higher.

image

### Study Communications
Team members can update study communications under Study editing, in the Actions button.

image


Study communications are visible on the main study page. Again, these communications are only visible for team members.

image

File: hawc/apps/study/views.py [hunk 1]
```diff
@@ -226,3 +237,46 @@ def get(self, request, *args, **kwargs):
...
+                Breadcrumb(
+                    name="Study " + str(study), url=reverse("study:detail", args=(study.id,))
+                ),
+            ],
+        )
+        return context
+
+    def get_object(self, queryset=None):
+        content_type = ContentType.objects.get(app_label="study", model="communication")
+        import pdb
```
Comment (rabstejnek) [inline hawc/apps/study/views.py:267]: Remove this

File: hawc/apps/assessment/views.py [hunk 2]
```diff
@@ -828,6 +848,41 @@ def get_context_data(self, **kwargs):
...
+        return context
+
+    def get_object(self, queryset=None):
+        content_type = ContentType.objects.get(app_label="assessment", model="communication")
+        obj, created = models.Communication.objects.get_or_create(
+            content_type=content_type, object_id=self.kwargs["pk"]
+        )
+        return obj
+
+    def post(self, request, *args, **kwargs):
```
Comment (shapiromatron) [inline hawc/apps/assessment/views.py:876]: try removing this and see if the complete lifecycle works; I don't think we need it.  The only thing you may need to change is to add a `get_success_url` method instead, and add the `reverse("assessment:detail", args=[kwargs["pk"]]`

File: hawc/apps/study/templates/study/study_detail.html [hunk 3]
```diff
@@ -21,7 +21,9 @@ <h2 class="d-inline-block">{{object}}</h2>
           {% if object.editable %}
             <a class="dropdown-item" href="{% url 'study:delete' object.pk %}">Delete study</a>
             <a class="dropdown-item" href="{% url 'study:attachment_create' object.pk %}">Add an attachment</a>
-
+            {% if user_is_team_member_or_higher %}
```
Comment (shapiromatron) [inline hawc/apps/study/templates/study/study_detail.html:24]: dont use custom here, use `obj.editable`; remove from view

File: hawc/apps/study/views.py [hunk 4]
```diff
@@ -226,3 +237,43 @@ def get(self, request, *args, **kwargs):
...
+        return context
+
+    def get_object(self, queryset=None):
+        content_type = ContentType.objects.get(app_label="study", model="communication")
+        obj, created = Communication.objects.get_or_create(
+            content_type=content_type, object_id=self.kwargs["pk"]
+        )
+        return obj
+
+    def post(self, request, *args, **kwargs):
```
Comment (shapiromatron) [inline hawc/apps/study/views.py:272]: again; try to remove the post as above

File: hawc/apps/study/views.py [hunk 5]
```diff
@@ -226,3 +237,43 @@ def get(self, request, *args, **kwargs):
...
+        content_type = ContentType.objects.get(app_label="study", model="study")
+        study = content_type.get_object_for_this_type(id=(self.object.object_id))
+        context["assessment"] = study.get_assessment()
+        context["breadcrumbs"] = Breadcrumb.build_crumbs(
+            self.request.user,
+            "Communication Update",
+            extras=[
+                Breadcrumb.from_object(context["assessment"]),
+                Breadcrumb(
+                    name="Study " + str(study), url=reverse("study:detail", args=(study.id,))
```
Comment (shapiromatron) [inline hawc/apps/study/views.py:259]: use f string here too

**Review by shapiromatron — CHANGES_REQUESTED** [review]
Looking good! Please see the requested edits. The key ones are use f-strings when you can, and try to use the default class-lifecycle as much as possible with the views.

**Review by shapiromatron — COMMENTED** [review]
Well done! Your changes look great. Really well done, now I think you know about 90% of the django + hawc view lifecycle.

Is it ok if I take this branch over now?  I'm going to make some changes that were out of scope of the original task; I'll ask you for a review when I'm complete?

**Review by shapiromatron — APPROVED** [review]
@casey1173 made a few changes to integrate with existing forms. Please review and let me know what you think.

Also, could you please it works for both create and update operations?

**Pull request merged**